universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+  @Transactional
+  public ProgramDefinition setBlockQuestions(
+      long programId, long blockDefinitionId, ImmutableList<QuestionDefinition> questionDefinitions)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setQuestionDefinitions(questionDefinitions)
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  @Transactional
+  public ProgramDefinition setBlockHidePredicate(
+      long programId, long blockDefinitionId, Predicate predicate)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setHidePredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  @Transactional
+  public ProgramDefinition setBlockOptionalPredicate(
+      long programId, long blockDefinitionId, Predicate predicate)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setOptionalPredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  int getBlockDefinitionIndex(ProgramDefinition programDefinition, Long blockDefinitionId)
+      throws ProgramBlockNotFoundException {
+    int index =
+        programDefinition.blockDefinitions().stream()
+            .map(b -> b.id())
+            .collect(ImmutableList.toImmutableList())
+            .indexOf(blockDefinitionId);
+    if (index == -1) {
+      throw new ProgramBlockNotFoundException(programDefinition.id(), blockDefinitionId);
+    }
```

Comment:
nit: please add a newline before and after the multiline `if` statement 🙏🏻

Reply:
done!